TRD: QC: Move root output

f3sch opened the pull request:
This patch allows to reproduce the actual online calibration which runs every 15 minutes. Every entry in the output tree represents a slot and contains the residual profiles and fit results.

Also I added the update for the configKeyValues, which I previously forgot...

Signed-off-by: Felix Schlepper <f3sch.git@outlook.com>

f3sch commented on Detectors/TRD/workflow/include/TRDWorkflow/VdAndExBCalibSpec.h:
```diff
@@ -76,11 +88,20 @@ class VdAndExBCalibDevice : public o2::framework::Task
     sendOutput(ec.outputs());
+    if (mOutFile && mOutTree) {
+      LOG(info) << "Writing calibration to file";
+      mOutFile->cd();
+      mOutTree->Write();
```
I am not sure if there is a better way to do this...

martenole replied:
similar here just a call to `mCalibrator->closeOutputFile()`. Besides that there is nothing wrong with the way you are writing

f3sch commented on Detectors/TRD/workflow/src/trd-calib-workflow.cxx:
```diff
@@ -37,6 +38,7 @@ void customize(std::vector<o2::framework::ConfigParamSpec>& workflowOptions)
 WorkflowSpec defineDataProcessing(ConfigContext const& configcontext)
 {
   auto enableRootInp = configcontext.options().get<bool>("enable-root-input");
+  o2::conf::ConfigurableParam::updateFromString(configcontext.options().get<std::string>("configKeyValues"));
```
I only tested the configurable params with the tracking workflow which already called this

martenole replied:
no problem, happens

f3sch commented on Detectors/TRD/calibration/src/CalibratorVdExB.cxx:
```diff
@@ -126,6 +126,18 @@ void CalibratorVdExB::initProcessing()
+    //mOutTree->Branch("residuals", &mFitFunctor.profiles);
+  }
+
+  // Retrieve configuration values
+  auto& params = TRDCalibParams::Instance();
+  mMinEntriesChamber = params.minEntriesChamber;
+  mMinEntriesTotal = params.minEntriesTotal;
```
I initial put this in the 'wrong' function. Programmatically it did matter and the result would have been the same.
But putting this here is saner.

f3sch commented on Detectors/TRD/calibration/src/CalibratorVdExB.cxx:
```diff
@@ -126,6 +126,18 @@ void CalibratorVdExB::initProcessing()
+  if (mOutTree) {
+    mOutTree->Branch("lorentzAngle", &mFitFunctor.laPreCorr);
+    mOutTree->Branch("vDrift", &mFitFunctor.vdPreCorr);
+    //mOutTree->Branch("residuals", &mFitFunctor.profiles);
```
ROOT does not support writing this _yet_.
Any idea how to get this in the tree without making a completely new object?

martenole replied:
Hm, `mFitFunctor.profiles` is an array of unique pointers to TProfile objects. I don't think that ROOT supports this. Maybe it would work with a loop, adding one branch per TProfile. Something like `mOutTree->Branch(fmt::format("residuals_{}", iDet), &(mFitFunctor.profiles[iDet]));`

martenole commented on Detectors/TRD/workflow/include/TRDWorkflow/VdAndExBCalibSpec.h:
```diff
@@ -47,7 +50,16 @@ class VdAndExBCalibDevice : public o2::framework::Task
     auto slotL = ic.options().get<uint32_t>("sec-per-slot");
     auto delay = ic.options().get<uint32_t>("max-delay");
-    mCalibrator = std::make_unique<o2::trd::CalibratorVdExB>(enableOutput);
+    if (enableOutput) {
```
I would prefer to move this block completely into `CalibratorVdExB.cxx`. And here inside the spec you could just call a function e.g. `mCalibrator->createOutputFile()` in case ROOT output is enabled. Then you could also leave the default constructor as it is now.
Inside the `CalibratorVdExB` you could then also switch to smart pointers which are preferred: `mOutFile = std::make_unique<TFile>("filename", "recreate")`

f3sch committed:
TRD: QC: Enable output for full configuration

This patch allows outputting the full vdrift exb calibration to a file.

For this two new function calls are defined which take care of the
creation and careful closing of the file.

Additionally, it adds a previously missed function call to update the
configKeyValues from cli args.

Signed-off-by: Felix Schlepper <f3sch.git@outlook.com>

martenole reviewed (approved):
Thanks! Will merge once the CI is through
